File: randomList.js
/*
* Based on "Fisher–Yates shuffle" algorithm https://en.wikipedia.org/wiki/Fisher–Yates_shuffle
* */
var RandomList = function(arr){
    'use strict';
    var cls = {};
    var i=0,rand,sel;

    cls.size = function(){
        return this.list.length;
    };

    cls.init = function(){
        cls.list = arr.slice(0);//Copying variable instead of reference.
    };

    cls.reset = function(){
        i=0,rand=0,sel='';
    };

    cls.getRand = function(){
        var size = cls.size();
        var min,max;
        min = i;
        max = size;
        if(min<max) {
            rand = Math.floor(min + Math.random() * (max - min));
            sel = this.list[rand];
            this.list[rand] = this.list[i];//shuffling value with "random key" & "current iterated key" within the range of pending iterations
            this.list[i] = sel;//shuffling value with "current iterated key" & "random key" within the range of pending iterations
            i++;
        }
        else{
            cls.reset();//Reset again. As, the reservoir is empty
            sel = cls.getRand();
        }

        return sel;
    };

    cls.init();//Initializing class

    return cls;
};

var arr = ["a", "b", "c"];
var randList = new RandomList(arr);
console.log(randList.getRand());
console.log(randList.getRand());
console.log(randList.getRand());
console.log("-----");
console.log(randList.getRand());
console.log(randList.getRand());
console.log(randList.getRand());
console.log("-----");
console.log(randList.getRand());
console.log(randList.getRand());
randList.list.push("d");
randList.list.push("e");
randList.list.push("f");
console.log(randList.getRand());
console.log(randList.getRand());
console.log("Original array unchanged: "+arr);

File: randomList.cs
/*
* Based on "Fisher–Yates shuffle" algorithm https://en.wikipedia.org/wiki/Fisher–Yates_shuffle
* */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace RandomListApplication
{
    class RandomList
    {
        private Int32 i, rand;
        private List<String> randList;
        private String sel;

        public String getRand()
        {
            var size = this.size();
            Int32 min = this.i;
            Int32 max = size;
            if (min<max)
            {
                Random rObj = new Random();
                this.rand = rObj.Next(min, max);
                this.sel = this.randList[this.rand];
                this.randList[this.rand] = this.randList[this.i];//shuffling value with "random key" & "current iterated key" within the range of pending iterations
                this.randList[this.i] = this.sel;//shuffling value with "current iterated key" & "random key" within the range of pending iterations
                this.i++;
            }
            else
            {
                this.reset();//Reset again. As, the reservoir is empty
                this.sel = this.getRand();
            }

            return this.sel;
        }

        public void reset()
        {
            this.i = 0;
            this.rand = 0;
            this.sel = "";
        }

        public Int32 size()
        {
            return this.randList.Count;
        }

        public RandomList(List<String> randList)
        {
            this.randList = randList;
        }
    }

    class ExecuteRandomList
    {
        static void Main(string[] args)
        {
            List<String> randList = new List<String>();
            randList.Add("a");
            randList.Add("b");
            randList.Add("c");
            RandomList randomList = new RandomList(randList);
            //Console.WriteLine(string.Join(",", randList));
            Console.WriteLine("Random 1: "+randomList.getRand());
            Console.WriteLine("Random 2: " + randomList.getRand());
            Console.WriteLine("Random 3: " + randomList.getRand());
            Console.WriteLine("Random 4: " + randomList.getRand());
            Console.ReadKey();
        }
    }
}
